### PR TITLE
Cols support - partly solved - first feedback

### PR DESCRIPTION
**Pull request by svenruppert**
This change is [Reviewable](https://reviewable.io/reviews/vaadin/board/46)

### CODE REVIEW

**Comment by rogozinds** [comment]
Reviewed 2 of 18 files at r2.
Review status: 2 of 16 files reviewed at latest revision, 8 unresolved discussions.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 79 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrKissMUG7sHxdkGZ9:-KjrKissMUG7sHxdkGZA:bqo58gl) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L79)):*
>         for (Component component : components) {
>             addComponent(component);
>         }
We don't need it, leave it as it was. Super.addComponents will handle this

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 92 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrKttn0dJ7SFFrGbZa:-KjrKttn0dJ7SFFrGbZb:b-h2xtnz) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L92)):*
>  //TODO check if Component is part ow Row Model
For whom is this todo? 

I think we can check that this row has this component, if yes execute all the code otherwise just ignore

so just wrap everything into
if(componenets.contains(c) {

}

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 122 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrLxV4ORjIO9KiY2Mx:-KjrLxV4ORjIO9KiY2My:bxh6am1) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L122)):*
>    public int usedColAmount() {
>         return getState().usedColAmount();
>     }
That's an API change, but I think this can be usefull, let's keep it for now.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 149 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrMGB0MB_OGOgL_-Bt:-KjrMGB0MB_OGOgL_-Bu:b-7ue89o) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L149)):*
>  public RowState getState() {
>         return (RowState) super.getState();
>     }
Why the method visibility has changed?

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 192 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-Kjr3AJvywpxKeDRqSjA:-Kjr3AJvywpxKeDRqSjB:bmsu0o2) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L192)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>     public void removeColsForComponent(Component component) {
>         setCols(component, 1);
>         //markAsDirty();
>     }
</details>
>  
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>     //    public void redraw() {
>     //        RowState state = getState();
>     //
>     //    }
</details>
Remove this method. It has confusing name, you cant remove cols for component. you can just set it to 1.
I think just having setCols(component, 1) is enough

---

*[integration-tests/pom.xml, line 268 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-Kjr2UQXBa80VI0P7OvD:-Kjr2UQXBa80VI0P7OvE:bnmrjmg) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/pom.xml#L268)):*
> ```XML
>             </build>
>         </profile>
>         
> ```

Why do we need docker? Can you remove it or move it to another PR.

---

*[integration-tests/src/main/java/com/vaadin/addon/board/testUI/BasicUI.java, line 4 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrN6y-Q7rnhGFi5r1w:-KjrN6y-Q7rnhGFi5r1x:b-ry7rwd) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/src/main/java/com/vaadin/addon/board/testUI/BasicUI.java#L4)):*
> ```Java
> 
> import com.vaadin.board.Board;
> import com.vaadin.board.Row;
> ```

I guess this is not needed, as it's the only change.

---

*[integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/AbstractParallelTest.java, line 17 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrNCHlzxiwp7ZGsHdF:-KjrNCHlzxiwp7ZGsHdG:b-el3mzs) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/AbstractParallelTest.java#L17)):*
> ```Java
> 
> import com.vaadin.testbench.annotations.BrowserConfiguration;
> import com.vaadin.testbench.annotations.RunLocally;
> ```

Remove runLocally.
Basically no tests in production should have this annotation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/board/46)*

**Comment by rogozinds** [comment]
Review status: 2 of 16 files reviewed at latest revision, 8 unresolved discussions.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 122 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrLxV4ORjIO9KiY2Mx:-KjrcUhWO7rk6WFM2tZZ:b74wobu) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L122)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

>    public int usedColAmount() {
>         return getState().usedColAmount();
>     }
That's an API change, but I think this can be usefull, let's keep it for now.
</blockquote></details>

We discuss this, it kinda hard to come up with a use case for this API.
One possible will be, when you want to add items to an existing board/row. 
Something like this:
you have a list of pictures, so you do something like
while(images.get) {
if(row.usedColAmount>=4) {
  row = new Row();
}
row.addComponent(image);
}
But it doesn't look like a typical use case, and it can be done smatter.
Let's keep the API minimal, because it's much easier to add new API later, than to remove existing one (we need to be compatable).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/board/46)*

**Comment by svenruppert** [comment]
Review status: 2 of 16 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 79 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrKissMUG7sHxdkGZ9:-Kk9abmnPpug2mvUwvJ_:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L79)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

>         for (Component component : components) {
>             addComponent(component);
>         }
We don't need it, leave it as it was. Super.addComponents will handle this
</blockquote></details>

Done.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 92 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrKttn0dJ7SFFrGbZa:-Kk9aYdO1G917stdsdqv:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L92)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

>  //TODO check if Component is part ow Row Model
For whom is this todo? 

I think we can check that this row has this component, if yes execute all the code otherwise just ignore

so just wrap everything into
if(componenets.contains(c) {

}
</blockquote></details>

Done.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 122 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrLxV4ORjIO9KiY2Mx:-Kk9aXoDC05gIuu0Rj9T:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L122)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

We discuss this, it kinda hard to come up with a use case for this API.
One possible will be, when you want to add items to an existing board/row. 
Something like this:
you have a list of pictures, so you do something like
while(images.get) {
if(row.usedColAmount>=4) {
  row = new Row();
}
row.addComponent(image);
}
But it doesn't look like a typical use case, and it can be done smatter.
Let's keep the API minimal, because it's much easier to add new API later, than to remove existing one (we need to be compatable).
</blockquote></details>

Done.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 149 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrMGB0MB_OGOgL_-Bt:-Kk9aWUp6ajGUy_sguim:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L149)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

>  public RowState getState() {
>         return (RowState) super.getState();
>     }
Why the method visibility has changed?
</blockquote></details>

Done.

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 192 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-Kjr3AJvywpxKeDRqSjA:-Kk9aiouRkRypPPalpuf:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L192)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

>     public void removeColsForComponent(Component component) {
>         setCols(component, 1);
>         //markAsDirty();
>     }
>  
>     //    public void redraw() {
>     //        RowState state = getState();
>     //
>     //    }
Remove this method. It has confusing name, you cant remove cols for component. you can just set it to 1.
I think just having setCols(component, 1) is enough
</blockquote></details>

Done.

---

*[integration-tests/pom.xml, line 268 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-Kjr2UQXBa80VI0P7OvD:-Kk9aktzOLw5ji8B_Gre:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/pom.xml#L268)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

Why do we need docker? Can you remove it or move it to another PR.
</blockquote></details>

Done.

---

*[integration-tests/src/main/java/com/vaadin/addon/board/testUI/BasicUI.java, line 4 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrN6y-Q7rnhGFi5r1w:-Kk9anrvUTSZAihpEpxU:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/src/main/java/com/vaadin/addon/board/testUI/BasicUI.java#L4)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

I guess this is not needed, as it's the only change.
</blockquote></details>

Done.

---

*[integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/AbstractParallelTest.java, line 17 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrNCHlzxiwp7ZGsHdF:-Kk9atFMRdUrlcQg9-5P:b-896fix) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/AbstractParallelTest.java#L17)):*
<details><summary><i>Previously, rogozinds (rogozinds) wrote…</i></summary><blockquote>

Remove runLocally.
Basically no tests in production should have this annotation.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/board/46)*

**Comment by rogozinds** [comment]
Reviewed 2 of 18 files at r2, 2 of 3 files at r3.
Review status: 2 of 15 files reviewed at latest revision, 13 unresolved discussions.

---

*[pom.xml, line 17 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-Kk9yRzMgBSV0ggQMZfc:-Kk9yRzMgBSV0ggQMZfd:b8rbjou) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/pom.xml#L17)):*
> ```XML
> 	</modules>
> ```

Remove from patch it's not a change

---

*[addon/src/main/java/com/vaadin/board/Board.java, line 68 at r4](https://reviewable.io:443/reviews/vaadin/board/46#-KkADWiRBHiAigvE70fW:-KkADWiRBHiAigvE70fX:b-7p7orw) ([raw file](https://github.com/vaadin/board/blob/4a65a8920d5bab152ed339433fd3c080bb3e5ff2/addon/src/main/java/com/vaadin/board/Board.java#L68)):*
>     public void removeCol(Row row, Component component){
>         row.removeColsForComponent(component);
>  
> //        markAsDirtyRecursive();
>         if(rows.contains(row)){
>         }
>     }
I think I"ve mentioned it before, that we do not want to have this API.
The user can just use row.setCols(Component, int).

Remove this. 

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 192 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-Kjr3AJvywpxKeDRqSjA:-KkAIstJFCch1TK9-4cY:b-z1j4lw) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/addon/src/main/java/com/vaadin/board/Row.java#L192)):*
<details><summary><i>Previously, svenruppert (Sven Ruppert) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Remove this method

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 94 at r3](https://reviewable.io:443/reviews/vaadin/board/46#-Kk9z98HqmGRTJO1zBOF:-Kk9z98HqmGRTJO1zBOG:boof201) ([raw file](https://github.com/vaadin/board/blob/2d53de0a22e956c24aee35443dff90dfd8b565be/addon/src/main/java/com/vaadin/board/Row.java#L94)):*
<details><summary><i alt="summary">Quoted 4 lines of code…</i></summary>
>     super.removeComponent(c);
>         //TODO check if Component is part ow Row Model
>         components.remove(c);
>         getState(true).cols.remove(c);
</details>

so just wrap everything into
if(componenets.contains(c) {

}

---

*[addon/src/main/java/com/vaadin/board/Row.java, line 152 at r3](https://reviewable.io:443/reviews/vaadin/board/46#-KkA-JEYzyldvHrisMjX:-KkA-JEYzyldvHrisMjY:b8z2600) ([raw file](https://github.com/vaadin/board/blob/2d53de0a22e956c24aee35443dff90dfd8b565be/addon/src/main/java/com/vaadin/board/Row.java#L152)):*
> cols_support

---

*[integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/ColRemoveIT.java, line 8 at r4](https://reviewable.io:443/reviews/vaadin/board/46#-KkAJiR2joGMDMdM3TIe:-KkAJiR2joGMDMdM3TIf:b8hyx63) ([raw file](https://github.com/vaadin/board/blob/4a65a8920d5bab152ed339433fd3c080bb3e5ff2/integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/ColRemoveIT.java#L8)):*
> ```Java
>  *
>  */
> public class ColRemoveIT extends AbstractParallelTest {
> ```

Implement the test.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/board/46)*

**Comment by rogozinds** [comment]
Review status: 2 of 15 files reviewed at latest revision, 15 unresolved discussions.

---

*[addon/src/main/java/com/vaadin/board/client/RowConnector.java, line 58 at r4](https://reviewable.io:443/reviews/vaadin/board/46#-KkAMDqiLF6ohmYjMLC-:-KkAMDqiLF6ohmYjMLC0:b-z5va06) ([raw file](https://github.com/vaadin/board/blob/4a65a8920d5bab152ed339433fd3c080bb3e5ff2/addon/src/main/java/com/vaadin/board/client/RowConnector.java#L58)):*
> ```Java
>                 .getWidget()
>                 .getElement()
>                 .setAttribute("board-cols", strValue);
> ```

so here add 
getWidget().redraw(getWidget().getElement());

---

*[addon/src/main/java/com/vaadin/board/client/RowWidget.java, line 11 at r4](https://reviewable.io:443/reviews/vaadin/board/46#-KkAMJ0YKuuvLZ7fAQHm:-KkAMJ0YKuuvLZ7fAQHn:b7k2cwi) ([raw file](https://github.com/vaadin/board/blob/4a65a8920d5bab152ed339433fd3c080bb3e5ff2/addon/src/main/java/com/vaadin/board/client/RowWidget.java#L11)):*
>  
>     public native void redraw()/*-{
>         this.redraw();
>     }-*/;
Change this to 
public native void redraw(Element elem)/*-{
  elem.redraw();
}-*/;

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/board/46)*

**Comment by rogozinds** [comment]
Reviewed 1 of 4 files at r5.
Review status: 3 of 16 files reviewed at latest revision, 9 unresolved discussions.

---

*[integration-tests/src/main/java/com/vaadin/addon/board/testUI/BasicUI.java, line 4 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrN6y-Q7rnhGFi5r1w:-KkAYpCwotYVhNLE1rF1:b-8dahjt) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/src/main/java/com/vaadin/addon/board/testUI/BasicUI.java#L4)):*
<details><summary><i>Previously, svenruppert (Sven Ruppert) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still there

---

*[integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/AbstractParallelTest.java, line 17 at r2](https://reviewable.io:443/reviews/vaadin/board/46#-KjrNCHlzxiwp7ZGsHdF:-KkAYunvNBP27iHxdySg:boof3vr) ([raw file](https://github.com/vaadin/board/blob/3f0d38d0db001804ac8c6815da270671ce3bdf17/integration-tests/src/test/java/com/vaadin/addon/board/testbenchtests/AbstractParallelTest.java#L17)):*
<details><summary><i>Previously, svenruppert (Sven Ruppert) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Still there.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/vaadin/board/46)*

**Comment by svenruppert** [comment]
https://github.com/vaadin/board/pull/50
